DllImportGenerator/designs/StructMarshalling.md

```diff
@@ -0,0 +1,127 @@
+- Limited type information in ref assemblies.
+  - In the ref assemblies generated by dotnet/runtime, we save space and prevent users from relying on private implementation details of structures by emitting limited information about their fields. Structures that have at least one non-object field are given a private `int` field, and structures that have at least one field that transitively contains an object are given one private `object`-typed field. As a result, we do not have full type information at code-generation time for any structures defined in the BCL when compiling a library that uses the ref assemblies.
+- Private reflection
+  - Even when we do have information about all of the fields, we can't emit code that references them if they are private, so we would have to emit unsafe code and calculate offsets manually to support marshaling them.
+  
+## Possible Solution A - Opt-In Structure Interop
+
+We've been working around another problem for a while in the runtime-integrated interop design: The user can use any type that is non-auto layout and has fields that can be marshaled in interop. This has lead to various issues where types that were not intended for interop usage became usable and then we couldn't update their behavior to be special-cased since users may have been relying on the generic behavior (`Span<T>`, `Vector<T>` come to mind).
+
+I propose an opt-in design where the owner of a struct has to explicitly opt-in to usage for interop. This enables our team to add special support as desired for various types such as `Span<T>` while also avoiding the private reflection and limited type information issues mentioned above.
+
+This design would use three attributes:
+
+```csharp
+
+public class GeneratedMarshallingAttribute : Attribute {}
+
+public class BlittableTypeAttribute : Attribute {}
+
+public class NativeStructMarshallingAttribute : Attribute
+{
+     public NativeStructMarshallingAttribute(Type nativeStructType, string managedToNativeMarshalMethod, string nativeToManagedMarshalMethod, string nativeFreeMethod) {}
+}
+
+```
+
+There are 2 usage mechanisms of these attributes.
+
+- Usage 1, Source-generated interop:
+
+The user can apply the `GeneratedMarshallingAttribute` to their structure `S`. The source generator will determine if the type is blittable. If it is blittable, the source generator will generate a partial definition and apply the `BlittableTypeAttribute` to the struct type `S`. Otherwise, it will generate a blittable representation of the struct and apply the `NativeStructMarshallingAttribute` and point it to the blittable representation, a method that marshals the managed value to the native value, a method that marshals a value of the native struct to a managed value, and a method that frees any native resources that were allocated for the native value. The blittable representation can either be generated as a separate top-level type or as a nested type on `S`.
```

Comment:
Should `NativeStructMarshallingAttribute` be allowed on classes as well? It does not look specific to structs.

Reply:
Is the thinking for integrating `ComWrappers` or similar for example?

Reply:
We can also allow it to be applied to classes similar to how we allow `StructLayoutAttribute` on classes to enable them for interop usage.

For ComWrappers support, we could either use a new attribute, or use Jan's idea for the "transparent struct" concept (if present, a property named `Value` on the native type would be the value we pass down instead of the native type) to implement marshalling via ComWrappers.

For example, something like the following would enable using ComWrappers to marshal a class via this model:

```
[NativeStructMarshalling(typeof(FooNative), ...)]
class Foo
{

}

class FooComWrappers : ComWrappers { ... }

class FooNative
{
      private static readonly FooComWrappers Wrapper = new FooComWrappers();
      public IntPtr Value { get; set; }
     ....
}

```


If we want to take this path to enable ComWrappers, I would suggest that we consider renaming `NativeStructMarshallingAttribute` to `NativeMarshallingAttribute`.